tests/lean/run/rewrite.lean

```diff
@@ -0,0 +1,64 @@
+/-!
+# Tests exercising basic behaviour of `rw`.
+-/
+
+def f (_ : Nat) : Nat := 0
+
+theorem t (n : Nat) : f n = f 0 := rfl
+theorem s (n m : Nat) : f n = f m := rfl
+
+attribute [irreducible] f
+
+-- Rewriting only affects the first occurrence.
+-- Providing an argument allows rewriting other occurrences.
+example : [f 1, f 2] = [f 0, f 0] := by
+  rw [t]
+  rw [t]
+  rw [t 2]
+
+-- We can rewrite backwards.
+example : f 1 = f 0 := by
+  rw [← t 2]
+  rw [t]
+  rw [t 2]
+
+-- Rewriting simultaneously replaces all occurrences of the first match.
+example : [f 1, f 2, f 1] = [f 0, f 0, f 0] := by
+  rw [t]
+  rw [t 2]
+
+-- Rewriting by lemmas with undetermined arguments creates additional goals.
+example : f 2 = f 3 := by
+  rewrite [s]
+  rfl
+
+-- These are often solved by the implicit `rfl` at the end of `rw`.
+example : f 2 = f 3 := by
+  rw [s]
+
+-- We can rewrite at multiple locations,
+-- specializing the arguments differently at each location.
+-- This behaviour was accidentally broken by
+-- https://github.com/leanprover/lean4/pull/2317
+-- and requiring a revert.
```

Comment:
Stylistically, should probably be a `/- ... -/` comment. It's a bit to long for line comments.